Add commitment burndown columns (#400)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/columns/commitmentdiscountunit.md

# Commitment Discount Unit

Commitment Discount Unit is the string value representing the provider-specified measurement unit that corresponds to [CommitmentDiscountConsumedQuantity](#commitmentdiscountconsumedquantity) and [CommitmentDiscountPurchaseQuantity](#commitmentdiscountpurchasequantity).

> CommitmentDiscountUnit MUST be present in the billing data when the provider supports [*commitment-based discounts*](#glossary:commitment-based-discount).
> 
> CommitmentDiscountUnit MUST not be null when [ChargeCategory](#chargecategory) is either *Purchase* or *Usage*, [PricingCategory](#pricingcategory) is *Committed*, and [CommitmentDiscountId](#commitmentdiscountid) is not null.
> 
> CommitmentDiscountUnit MUST be null in all other cases.

Reviewer: ```suggestion
The CommitmentDiscountUnit column adheres to the following requirements:

 * CommitmentDiscountUnit MUST be present in the billing data when the provider supports [*commitment-based discounts*](#glossary:commitment-based-discount).
 * CommitmentDiscountUnit MUST be of type String, and the units of measure used in CommitmentDiscountUnit SHOULD adhere to the values and format requirements specified in the [UnitFormat](#unitformat) attribute.
 * CommitmentDiscountUnit MUST NOT be null if [ChargeCategory](#chargecategory) is "Usage" or *Purchase* and [*CommitmentDiscountId*](#commitmentdiscountid) is not null, unless [ChargeClass](#chargeclass) is "Correction".
 * CommitmentDiscountUnit MAY be null if ChargeCategory is "Usage" or "Purchase" and ChargeClass is "Correction".
 * CommitmentDiscountUnit MUST be null in all other cases.
```

# Commitment Discount Unit

> Commitment Discount Unit is the string value representing the provider-specified measurement unit that corresponds to [Commitment Discount Quantity](#commitmentdiscountconsumedquantity) and [Commitment Discount Purchased Quantity](#commitmentdiscountpurchasedquantity) of a [*commitment discount*](#glossary:commitment-discount). The CommitmentDiscountUnit column is only applicable to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount).

Reviewer 2: This makes sense for usage commitments. What about spend commitments? This doesn't convey what unit I should use for those.

Author: You should hours but it's up to you

Reviewer 3: FOCUS can't dictate the units right. Spend commitments I would think will use some sort of currency as the unit.

The CommitmentDiscountUnit column adheres to the following requirements:

* CommitmentDiscountUnit MUST be present in the billing data when the provider supports [*commitment-based discounts*](#glossary:commitment-based-discount).

> * CommitmentDiscountUnit MUST be of type String, and the units of measure used in CommitmentDiscountUnit SHOULD adhere to the values and format requirements specified in the [UnitFormat](#unitformat) attribute.

Author: This is an interesting question, but since the same clause exists for ConsumedUnit and PricingUnit, I'm going to keep as is until a more holistic approach is offered.

> * CommitmentDiscountUnit MUST be of type String, and the units of measure used in CommitmentDiscountUnit SHOULD adhere to the values and format requirements specified in the [UnitFormat](#unitformat) attribute.

Author: Same comment as above not taking up this effort not

> * CommitmentDiscountUnit MUST be of type String, and the units of measure used in CommitmentDiscountUnit SHOULD adhere to the values and format requirements specified in the [UnitFormat](#unitformat) attribute.

Reviewer 2: Shouldn't we mandate that this use the same value as ConsumedUnit for usage commitments? Should we mandate that it be the BillingCurrency for spend commitments? If not, what would it be for spend commitments?
```suggestion
* CommitmentDiscountUnit MUST be of type String, and the units of measure used in CommitmentDiscountUnit SHOULD adhere to the values and format requirements specified in the [UnitFormat](#unitformat) attribute.
* When [ChargeCategory](#chargecategory) is "Usage" and [CommitmentDiscountCategory](#commitmentdiscountcategory) is "Usage", CommitmentDiscountUnit MUST use the same value as [ConsumedUnit](#consumedunit).
* When [ChargeCategory](#chargecategory) is "Usage" and [CommitmentDiscountCategory](#commitmentdiscountcategory) is "Spend", CommitmentDiscountUnit MUST use the same value as [BillingCurrency](#billingcurrency).
```

Author: With size flex, "Normalized Hours" are neither the `ConsumedUnit` or the `BillingCurrency`

Reviewer 2: Does the Unit Format attribute have an affordance for "Normalized"? That might be worth calling out there. Either way, if we need to account for normalization, we can easily add that to the requirements. Then we can say it SHOULD be "Normalized Foo" when normalization is applied and matching ConsumedUnit when it's not normalized.

Reviewer: Follow-up on suggestion by @flanakin:

#### Commitment Discount Category: Spend

- This allows us to limit Spend to BillingCurrency and exclude non-currency consumable units (NCCUs), as @rileyjenk calls them? Providers can still support NCCUs and use these columns if all other commitment discount requirements are met, but they need to convert NCCUs into BillingCurrencies.
- If we decide to impose this, CommitmentDiscountConsumedQuantity would align with EffectiveCost in the case of Usage, and CommitmentDiscountPurchasedQuantity would align with BilledCost in case of Purchase. I’m not suggesting that explicitly stating this adds any value; I’m simply pointing this out.

#### Commitment Discount Category: Usage

- Normalized Hours are essentially Hours, and by calling them "Normalized," we imply that the value in CommitmentDiscountConsumedQuantity is normalized? I would prefer to use the same value for both CommitmentDiscountUnit and ConsumedUnit and specify that CommitmentDiscountConsumedQuantity includes the size normalization factor (size normalization factor should be added to the glossary).
- Using the same value for both CommitmentDiscountUnit and ConsumedUnit would help practitioners easily detect charges with normalization factors by comparing CommitmentDiscountConsumedQuantity and ConsumedQuantity, and even calculate the factor by dividing these quantities.

- **NOTE:** Keep in mind that we have specified that for Unused Usage-related records, CommitmentDiscountUnit MUST NOT be null, but ConsumedUnit MUST be null. If we choose this approach, we need to **either rephrase the requirement or revert to having ConsumedQuantity and ConsumedUnit not be null in the case of Unused Usage**.

> * CommitmentDiscountUnit MUST NOT be null when [*CommitmentDiscountId*](#commitmentdiscountid) is NOT null and *ChargeClass* is not "Correction".

Author: Actually, all `is NOT` needs to be `is not`, so fixing this way.  Good catch though ... 
<img width="1280" alt="image" src="https://github.com/user-attachments/assets/4bcc1e8c-9adb-40fe-9d47-4e33f5d4cf8b">

Commitment Discount Unit represents a provider-specified measurement unit indicating how a provider measures [Commitment Discount Quantity](#commitmentdiscountquantity) of a [*commitment discount*](#glossary:commitment-discount). The CommitmentDiscountUnit column is only applicable to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount).

The CommitmentDiscountUnit column adheres to the following requirements:

* CommitmentDiscountUnit MUST be present in a FOCUS dataset when the provider supports [*commitment discounts*](#glossary:commitment-discount).
* CommitmentDiscountUnit MUST be of type String, and the units of measure used in CommitmentDiscountUnit SHOULD adhere to the values and format requirements specified in the [UnitFormat](#unitformat) attribute.
* The CommitmentDiscountUnit MUST be the same across all *rows* where *CommitmentDiscountQuantity* has the same [*CommitmentDiscountId*](#commitmentdiscountid).
* CommitmentDiscountUnit MUST NOT be null when *CommitmentDiscountId* is not null.
* CommitmentDiscountUnit MUST be null in all other cases.

In cases where the *CommitmentDiscountQuantity* are present, the following applies:

> * The CommitmentDiscountUnit MUST be the unit measured by the *commitment discount*.

Reviewer: ```suggestion
* The CommitmentDiscountUnit MUST represent the unit used to measure the *commitment discount*.
```

# Commitment Discount Unit

> Commitment Discount Unit is the string value representing the provider-specified measurement unit that corresponds to [Commitment Discount Quantity](#commitmentdiscountquantity) of a [*commitment discount*](#glossary:commitment-discount). The CommitmentDiscountUnit column is only applicable to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount).

Reviewer 3: ```suggestion
Commitment Discount Unit represents a provider-specified measurement unit indicating how a provider measures [Commitment Discount Quantity](#commitmentdiscountquantity) of a [*commitment discount*](#glossary:commitment-discount). The CommitmentDiscountUnit column is only applicable to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount).
```

The CommitmentDiscountUnit column adheres to the following requirements:

> * CommitmentDiscountUnit MUST be present in the billing data when the provider supports [*commitment discounts*](#glossary:commitment-discount).

Reviewer 3: ```suggestion
* CommitmentDiscountUnit MUST be present in a FOCUS dataset when the provider supports [*commitment discounts*](#glossary:commitment-discount).
```

* CommitmentDiscountUnit MUST be of type String, and the units of measure used in CommitmentDiscountUnit SHOULD adhere to the values and format requirements specified in the [UnitFormat](#unitformat) attribute.
* The CommitmentDiscountUnit MUST be the same across all *rows* where *CommitmentDiscountQuantity* has the same [*CommitmentDiscountId*](#commitmentdiscountid).

> * CommitmentDiscountUnit MUST NOT be null when *CommitmentDiscountId* is not null.

Reviewer 3: when its a correction we're not asking for this definition as a MUST right? MUST when not a correction and SHOULD otherwise? @ijurica

Reviewer: Correct.
As I recall, we planned to create a Git issue to address this as part of our Consistency review—specifically, to review and align how we specify Corrections across the specification.

In other columns (costs, quantities, units, etc.), we **allow null values for Corrections**. Additionally, we **didn’t address negative values explicitly**, even though it’s clear that quantities and costs can be negative only in the case of Correction charges and must be non-negative for non-Correction charges. . Please discuss (briefly), make a decision, and apply it consistently across the specification.

* CommitmentDiscountUnit MUST be null in all other cases.

In cases where the *CommitmentDiscountQuantity* are present, the following applies:

> * When the *commitment discount* commits to a usage amount over a term, the CommitmentDiscountUnit SHOULD be "Hours".
> * When the *commitment discount* commits to a usage amount over a term where size-flexibility is applied, the CommitmentDiscountUnit SHOULD be "Normalized Hours".

Reviewer 3: ```suggestion
* When the [*CommitmentDiscountCategory*](#commitmentdiscountcategory) is 'Usage', where size-flexibility is applicable for the commitment discount, the CommitmentDiscountUnit SHOULD be "Normalized Hours". If size-flexibility is not applicable, CommitmentDiscountUnit SHOULD be "Hours".
```

> * When the *commitment discount* commits to a spend amount over a term, the CommitmentDiscountUnit SHOULD match the [*BillingCurrency*](#billingcurrency).

Reviewer 3: ```suggestion
* When *CommitmentDiscountCategory* is 'Spend', the CommitmentDiscountUnit SHOULD match the [*BillingCurrency*](#billingcurrency).
```

## Column ID

CommitmentDiscountUnit

## Display Name

Commitment Discount Unit

## Description

> The string value representing the provider-specified measurement unit that corresponds to *CommitmentDiscountQuantity* and *CommitmentDiscountPurchasedQuantity* of a *commitment discount*.

Reviewer 3: ```suggestion
The provider-specified measurement unit indicating how a provider measures [Commitment Discount Quantity](#commitmentdiscountquantity) of a [*commitment discount*](#glossary:commitment-discount).
```
